feat: Update example to match current EKS blueprints, remove experimental feature to allow users to use Terraform 1.3+

bryantbiggs opened the pull request:
- Remove `module_variable_optional_attrs` which is removed in v1.3.0 of Terraform; this currently is holding back EKS Blueprints from being able to suport - https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/988
	- Closes #7
- General updates to better align with current EKS Blueprints project - see notes below and throughout changeset

bryantbiggs commented on blueprints/getting-started/main.tf:
```diff
@@ -1,80 +1,78 @@
+################################################################################
+
+module "eks_blueprints" {
+  source = "github.com/aws-ia/terraform-aws-eks-blueprints?ref=v4.12.1"
```
pin version for repeatability/stability

bryantbiggs replied:
removing pin due to cyclical conflict

bryantbiggs commented on blueprints/getting-started/main.tf:
```diff
@@ -1,80 +1,78 @@
 module "eks_blueprint_addons" {
-  source = "github.com/aws-ia/terraform-aws-eks-blueprints//modules/kubernetes-addons"
-  #source = "../../../aws-eks-accelerator-for-terraform/modules/kubernetes-addons/"
+  source = "github.com/aws-ia/terraform-aws-eks-blueprints//modules/kubernetes-addons?ref=v4.12.1"
```
pin version for repeatability/stability

bryantbiggs commented on blueprints/getting-started/main.tf:
```diff
@@ -83,12 +81,48 @@ module "eks_blueprint_addons" {
+    "kubernetes.io/role/internal-elb" = 1
   }
 
-  depends_on = [
```
We shouldn't depend on other modules and instead map the dependency ordering by passing outputs to the modules inputs for implicit dependency mapping. 

Using `depends_on` here will cause disruptions for users, specifically around IAM policies which cause service disruption - ref https://github.com/hashicorp/terraform/issues/30340#issuecomment-1010202582

bryantbiggs commented on locals.tf:
```diff
@@ -1,52 +0,0 @@
-locals {
-  default_helm_values = [
-    templatefile("${path.module}/vault-config.tftpl.yml", {})
```
`file()` works here when we aren't templating and simply reading in a file

ksatirli replied:
It does - I think originally, we planned on templatizing this more, but then decided not to.

Thanks for cleaning this up!

bryantbiggs commented on locals.tf:
```diff
@@ -1,52 +0,0 @@
-    wait_for_jobs              = false
-    description                = "Helm chart to install Vault and other associated components"
-    verify                     = false
-    keyring                    = ""
```
we can take advantage of things like `lookup()` or `try()` to first attempt to get the value from `var.helm_config` if the user has provided one, otherwise fall back to a default value. When falling back to a default value and we don't have an opinion on an optional argument, we should default to `null` to indicate as such and leave it up to the resource's defaults

bryantbiggs commented on main.tf:
```diff
@@ -1,63 +1,66 @@
+  namespace        = try(var.helm_config.namespace, "vault")
+  create_namespace = try(var.helm_config.create_namespace, true)
+  description      = try(var.helm_config.description, null)
+  chart            = "vault"
```
this should be hardcoded, its not configurable nor optional

bryantbiggs commented on terraform.tf:
```diff
@@ -1,7 +1,10 @@
+  required_version = ">= 1.0"
+
+  required_providers {
+    helm = {
```
since we are using the helm provider for `helm_release`, we should declare that the helm provider is required - https://www.terraform.io/language/providers/requirements#requiring-providers

bryantbiggs commented on variables.tf:
```diff
@@ -1,34 +1,11 @@
-variable "addon_context" {
```
https://github.com/terraform-linters/tflint is great for finding un-used variables/locals and other common conventions

ksatirli replied:
I dug into this and am grateful for you spotting this and reporting it!

The Superlinter [config](https://github.com/hashicorp/terraform-aws-hashicorp-vault-eks-addon/blob/main/.github/workflows/superlinter.yml#L27) for this repository has `VALIDATE_TERRAFORM` enabled and that used to take care of `tflint`. 

As it turns out, in `v4` of Superlinter, `VALIDATE_TERRAFORM_TFLINT` is the correct way for `tflint` to be enabled, as more granular control was added.

I've fixed this in the long-running PR of https://github.com/hashicorp/terraform-aws-hashicorp-vault-eks-addon/pull/3 and also merged this after brief discussion.

For other readers: #3 has a couple of `tflint` warnings that are fixed by this PR (#8)

bryantbiggs commented:
@ksatirli would you mind taking a look when you get a chance so we can resolve the upstream issue related to Terraform v1.3+

ksatirli commented on blueprints/getting-started/main.tf:
```diff
@@ -1,80 +1,78 @@
-# See https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/region
-data "aws_region" "current" {}
```
Great sleuthing and fix!

ksatirli commented:
@bryantbiggs these look great on first look, I'll take an in-depth look in the morning. 

Excited about the changes you're proposing (with `vault_namespace` being the _only_ thing that immediately stood out as a case of [unfortunate naming](https://developer.hashicorp.com/vault/docs/enterprise/namespaces))

bryantbiggs commented:
Thanks @ksatirli - I've updated to change `vault_namespace` to simply `namespace` in the example. I also removed the version pin due to a cyclical conflict between external modules and the Blueprints repository/version

ksatirli commented:
> Thanks @ksatirli - I've updated to change `vault_namespace` to simply `namespace` in the example. I also removed the version pin due to a cyclical conflict between external modules and the Blueprints repository/version

Thanks - These changes look great!

I vaguely remember there being a reason to _not_ pin the Blueprints module (which is - as you point out - an anti-pattern), but I can't find the actual _why_.

bryantbiggs commented:
Yes, I think the way that we have both the module definition for the `helm_addon` and the module implementation in the same repository coupled with the fact that we have external addons like Vault that also point to this addon and are referenced from within the same project, it creates a nasty cyclical loop. I will have a chat with the team to see if we can do something about moving the `helm_addon` to its own module repo to break this loop and then we can utilize versions better

ksatirli commented:
@bryantbiggs I've tested this locally with `1.3.1` and found it working. 

I've enabled `tflint` via #3, which caused a small merge conflict in `blueprints/getting-started/main.tf` - happy to resolve this or let you be the sole committer of this PR.

ksatirli commented:
Code looks great @bryantbiggs - the `terraform-docs` Action will be fixed in an upcoming PR.

Unless you need any more changes, I'm good with merging this.

ksatirli commented:
Time to :shipit:.

Thank you once again, @bryantbiggs
